### DISCUSSION TITLE
Plans for future development? Project concept?

### DISCUSSION THREAD

**Discussion by roman-dvorak** [inventree/InvenTree#4920] [Ideas]
Hello,

for several years we have been developing a similar solution for inventory tracking - [OpenIntranet](https://github.com/UniversalScientificTechnologies/OpenIntranet). Unfortunately, we have been mostly developing it on our own and haven't been able to build a community around it (possibly due to the lack of quality documentation, texts in Czech language, difficult first setup etc.). I have been testing InvenTree for some time now, and I like its concept. When we started developing our OpenIntranet, we wanted to use a low-level web framework, so we chose Python-Tornado and a NoSQL MongoDB database. This decision has brought both advantages and disadvantages (which could be a lengthy discussion). The OpenIntrant system is designed to be maximally user-friendly, simple with fast responsiveness.

Currently, our system handles several areas of functionality:
  *  Warehouse management
   * Time tracking and payroll management
   * Advanced tools for inventory management
   * Advanced tools for electronic component manufacturing (PCB assembly, completion)

In our further plans, we intended to implement order management and automate purchasing from component suppliers.

After testing InvenTree, I'm considering whether we should transfer a certain part of the functionality/agenda to InvenTree. There are aspects missing in all parts of our system, and if possible, I would like to discuss whether there is interest in such collaboration or not. I/we are able to develop/improve some parts of InvenTree. 

## Warehouse management

We have a similar organization of our warehouse, structured in a tree-like manner. What I find missing in InvenTree is that we often work in multiple locations as a team of collaborating individuals. Therefore, our inventory is physically distributed across multiple locations.

![obrazek](https://github.com/inventree/InvenTree/assets/5196729/d24bbfcf-da68-4d4f-8759-dab9a3da9b56)

That's why for each item, we always specify the quantity in stock at a current location, as well as in other locations. And whenever any interaction occurs, it is based on the inventory of the specific location where the user is currently located.

![obrazek](https://github.com/inventree/InvenTree/assets/5196729/49fd7496-805d-4eda-9ccc-78afde47a016)

> :question:  I have repeatedly noticed in InvenTree that during manufacturing, I can use components from anywhere, even if they are not readily available at the moment. Is there any plan to solve this somehow?
> I could imagine having a setting for the "Actual stock root," from which subsequent operations would be based. 

> :question: I couldn't find a way to access the history of operations for a stock items in InvenTree. Are such data available or not?

In our system, we obtaining most of the information about components (description, specifications, images) from supplier APIs (Mouser, TME, Farnell). Regarding this capability, I have contacted the author of this plugin: https://github.com/SergeoLacruz/inventree-supplier-plugin/issues/3

## Identificators
I generally have an problem with the identifiers of individual items (such as inventory, manufacturing, suppliers, items, etc.) being simple numbers. Because in case of any error, they can be easily confused/swapped (for example), especially when multiple InvenTree installations are involved. The code from one installation may work with the code from another instance, but it can have completely different meanings.

Our solution was to use [ObjectID](https://www.mongodb.com/docs/manual/reference/bson-types/#objectid) for such identifiers, which is a globally unique code. Therefore, within the system (even across instances), the same code cannot exist with a different meaning.

> :question:  This is a very important point for me, but I still don't know if it has any solution. Do you have any opinion on this?

Thanks to this, we can easily have a unique barcode (DataMatrix) on each object in the warehouse (bin location, item). This code on sticker refers to a specific "package," how we call it. In InvenTree, the equivalent term is "Part stock."

![obrazek](https://github.com/inventree/InvenTree/assets/5196729/bbf1601b-5c85-411b-b146-72d574b0accd)

## Manufacturing

The manufacturing agenda is currently very important for us. It looks like that we generate an XML netlist from KiCad (fully automted with github actions), which is then uploaded to the intranet. The intranet retrieves the Bill of Materials (BOM) from the XML and links it to the items in the warehouse based on the ObjectID component identifier. (This is where the importance of having globally unique identifiers becomes evident.)

Various overviews or PDF assembly instructions are then created from this data, including a list of components with their locations in the warehouse and their available quantities.

> :question:  Regarding this, I have a question: Are such overviews desirable for you, and would you be open to implement them into the manufacturing part? Additionally, I would like to further integrate it with an [interactive HTML BOM](https://openscopeproject.org/InteractiveHtmlBomDemo/) that would facilitate the assembly process.

![obrazek](https://github.com/inventree/InvenTree/assets/5196729/3f2ff888-a7fb-469d-8171-684267a15f2f)
![obrazek](https://github.com/inventree/InvenTree/assets/5196729/1d31d2b8-e345-4e94-858f-82e9c24fb709)

That's my currently the most important information. There would definitely be more questions, and I apologize for the lengthy text. I look forward to your responses and potential cooperation.

**Comment by matmair** [comment]
Hi there @roman-dvorak ! The opinion below is only my personal view on things.

Looks interesting but other than concepts we can not take/integrate anything from your project as it is licensed under GPL and InvenTree is MIT.
You could integrate via the REST API, the Python plugin or the plugin system but the other way around might be difficult. This is a bit sad but such is the nature of permissive vs copyleft licenses. That can not really be changed after the fact and InvenTree being permissive is - at least for me personally - one of the key points that lead to my contributing.

It looks like the 2 projects came to similar conclusions with different architectures - which is interesting and validating.

**Reply by roman-dvorak** [reply]
The OpenIntranet license is one thing, and we can work with it in some way since OpenIntranet is fully under our control. 

I think that the approach and methods used inside are much more important than the implementation itself. Considering that OpenIntranet uses a NoSQL database and Tornado, a own implementation with Django would be quite different.

For some actions, it is beneficial to have API level interaction between tools. However, there are things that are more suitable to be implemented within a single system.

**Reply by matmair** [reply]
Would you like to switch to InvenTree and have the changes implemented by the org/ implement them? English is not my first language so I might not fully understand your suggestion. What is the goal of the cooperation you are writing about/how would you imagine it working?

**Reply by roman-dvorak** [reply]
Yes, I'm currently in the process of finding out what the general features and development plans for InvenTree are. According on that, I want to decide whether we could implement these features directly into InvenTree.

Yes, you understand correctly. I'm considering moving part of our agenda to InvenTree. For example stock and manufacturing management. 

How would I imagine it? I would imagine it as follows: We have our certain requirements/ideas about how the warehouse and production system should work. Based on that, we would consult with you and implement some parts ourselves (and provide them for community - Of course! :) ). For other issues/request, we would provide our own ideas or suggestions.

PS: Don't hesitate to ask me for further explanation of things. I'm also not a native, my English isn't perfect.

**Comment by SchrodingersGat** [comment]
Hey @roman-dvorak thanks for reaching out, looks like an interesting project with similar goals :)

I'll respond to your questions inline below:

>  I have repeatedly noticed in InvenTree that during manufacturing, I can use components from anywhere, even if they are not readily available at the moment. Is there any plan to solve this somehow?

It is true that InvenTree is designed as a "global" warehouse system - although we do support a multiple tree structure for warehouse / stock locations, so there's nothing preventing us from improving on this. We could (for example) allocate each user to a particular top-level warehouse location, and prevent them from operating in different locations.

> I could imagine having a setting for the "Actual stock root," from which subsequent operations would be based.

See above!

Note that for build orders, you can select a "take stock from" location, which is somewhat of a solution for this issue.

> Our solution was to use [ObjectID](https://www.mongodb.com/docs/manual/reference/bson-types/#objectid) for such identifiers, which is a globally unique code. Therefore, within the system (even across instances), the same code cannot exist with a different meaning.

Seems like a reasonable approach, and we could look to extend our barcoding system with something like this. Currently, two barcodes from multiple InvenTree instances could certainly conflict, although that seems like a minimal issue given how many other outstanding issues we have.

> The manufacturing agenda is currently very important for us. It looks like that we generate an XML netlist from KiCad (fully automted with github actions), which is then uploaded to the intranet. The intranet retrieves the Bill of Materials (BOM) from the XML and links it to the items in the warehouse based on the ObjectID component identifier.

Integration with other tools (specifically such as KiCad) is a big goal for us. interactive html bom would be a great one to integrate too. There is some work towards this:

- https://github.com/inventree/InvenTree/discussions/4133 - this is more for extracting data *from* InvenTree and *into* KiCad

But the reverse operation (KiCad data into InvenTree) is very desirable too and a personal goal for me. 

We have a mature plugin system, which allows such integrations. Check out [inventree-wireviz](https://github.com/inventree/inventree-wireviz) (work in progress) for integration with the wireviz harness design tool.

All of your ideas here have merit and I'd love to see InvenTree implement similar things in some way. At the moment our biggest bottleneck (by *far*) is developer time. The project has become very popular, so a lot of issues, questions, ideas (etc) to support.

If you would like to contribute ideas / **code** to the project, I'd love to support that. Certainly it looks like you have a great deal of experience and ideas :)

**Reply by roman-dvorak** [reply]
Thank you for your reply @SchrodingersGat

> > Our solution was to use [ObjectID](https://www.mongodb.com/docs/manual/reference/bson-types/#objectid) for such identifiers, which is a globally unique code. Therefore, within the system (even across instances), the same code cannot exist with a different meaning.
> 
> Seems like a reasonable approach, and we could look to extend our barcoding system with something like this. Currently, two barcodes from multiple InvenTree instances could certainly conflict, although that seems like a minimal issue given how many other outstanding issues we have.

I have been searching for a while, and I found that both Django and PostgreSQL natively support UUIDs ([postgress](https://www.postgresql.org/docs/current/datatype-uuid.html), [Django](https://www.geeksforgeeks.org/uuidfield-django-models/)), which are globally unique identification codes. This is equivalent to our [ObjectID](https://www.mongodb.com/docs/manual/reference/bson-types/#objectid). It seems that this could potentially solve this (our) problem. Currently, I have no idea what implementing such a change would mean for InvenTree.


> We have a mature plugin system, which allows such integrations. Check out [inventree-wireviz](https://github.com/inventree/inventree-wireviz) (work in progress) for integration with the wireviz harness design tool.

We have saw this plugin several times on GitHub. I know the WireViz tool (I have used it), but from the description, I cannot quite understend how it is used inside InvenTree. But I assume that it is a temporary issue :)

> Integration with other tools (specifically such as KiCad) is a big goal for us. interactive html bom would be a great one to integrate too.

Great! I propose that I create a separate thread for this issue, where I'll try to describe the proposed (ideal?) solution for the coperation between KiCAD and the Stock management system.

**Comment by SchrodingersGat** [comment]
@roman-dvorak looking forward to your contributions and ideas to make Inventree better for everyone :)

**Comment by roman-dvorak** [comment]
One quick observation:

Compared to our OpenIntranet, it is much slower. Rendering of single page, from the first request to the state when the page is usable, takes much more time. Also, asynchronous loading (Which we also have, but used in different ways) causes the page to redraw during operation time.

For example, the loading of the left menu is done after the rendering of main content.
I don't have an estimate of what causes this, but it might be due to the deep nesting of individual python methods and their branching, probably due to use of Django. Another factor could be some database aggregations. OpenIntranet (thanks to its NoSQL database) doesn't require as many links between collections. 

It is used/compared on the same hardware.  I dont have configured caching using Redis - I hope that will improve the situation. After configuration, I will let you know how the situation were changed. 

HW: 
> 8 cores of AMD EPYC 7H12
> RAM 4 GB
> OS: NixOS, 
> Installation: Production docker-compose


Question:
Is it okay if I create a set of [FR] issues for some ideas that may not be immediately solved?

**Reply by matmair** [reply]
That is mainly because of the way the interface is built. We use a lot of templates and vanilla js - so a lot of data has to be transferred per call. There are some experiments with using React - which enables a far faster UI experience. The API itself is pretty fast - maybe try interacting with it or the native apps to get a feel for the possible speed.

Feel free to create FRs. Please look into the issues (open and closed) before creating new ones and try to fill out the template as much as possible as that helps with the proper scoping of development.
Edit: Also try to keep FRs to single, actionable items
